src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -41,29 +41,105 @@ static const cmatrix_t one_measure =
+uint_t reorder_qubits(const reg_t qubits, uint_t index);
+
 //--------------------------------------------------------------------------
 // Function name: reverse_all_bits
 // Description: The ordering of the amplitudes in the statevector in this module is 
 //    000, 001, 010, 011, 100, 101, 110, 111.
 //    The ordering of the amplitudes in the statevector in Qasm in general is 
 //    000, 100, 010, 110, 001, 101, 011, 111.
 //    This function converts the statevector from one representation to the other.
+//    This is a special case of reorder_qubits
 // Input: the input statevector and the number of qubits
-// Returns: the statevector in reverse order
+// Output: the statevector in reverse order
 //----------------------------------------------------------------	
-cvector_t reverse_all_bits(const cvector_t& statevector, uint_t num_qubits);
+template <class T>
+std::vector<T> reverse_all_bits(const std::vector<T>& statevector, uint_t num_qubits);
 uint_t reverse_bits(uint_t num, uint_t len);
-  std::vector<uint_t> calc_new_indexes(std::vector<uint_t> indexes);
+ std::vector<uint_t> calc_new_indexes(std::vector<uint_t> indexes);
 
 // The following two functions are helper functions used by 
 // initialize_from_statevector
 cmatrix_t reshape_matrix(cmatrix_t input_matrix);
 cmatrix_t mul_matrix_by_lambda(const cmatrix_t &mat,
 		               const rvector_t &lambda);
 
+
 //------------------------------------------------------------------------
 // local function implementations
 //------------------------------------------------------------------------
+template <class T>
+void reorder_all_qubits(const std::vector<T>& orig_probvector, 
+			reg_t qubits,
+			std::vector<T>& new_probvector) {
+  uint_t new_index;
+  uint_t length = 1ULL << qubits.size();   // length = pow(2, num_qubits)
+  // if qubits are [k0, k1,...,kn], move them to [0, 1, .. , n], but preserve relative
+  // ordering
+  reg_t squeezed_qubits(qubits.size());
+  std::vector<uint_t> sorted_qubits;
+  for (uint_t index : qubits) {
+    sorted_qubits.push_back(index);
+  }
+  sort(sorted_qubits.begin(), sorted_qubits.end());
+  for (uint_t i=0; i<qubits.size(); i++) {
+    for (uint_t j=0; j<sorted_qubits.size(); j++) {
+      if (qubits[i] == sorted_qubits[j]) {
+	squeezed_qubits[i] = j;
+	break;
+      } 
+    }    
+  }
+
+  for (uint_t i=0; i < length; i++) {
+    new_index = reorder_qubits(squeezed_qubits, i);
+    new_probvector[new_index] = orig_probvector[i];
+  } 
+}
+
+uint_t reorder_qubits(const reg_t qubits, uint_t index) {
+  uint_t new_index = 0;
+
+  int_t current_pos = 0, current_val = 0, new_pos = 0, shift =0;
+  uint_t num_qubits = qubits.size();
+  for (uint_t i=0; i<num_qubits; i++) {
+    current_pos = num_qubits-1-qubits[i];
+    current_val = 0x1 << current_pos;
+    new_pos = num_qubits-1-i;
+    shift = new_pos - current_pos;
+    if (index & current_val) {
+      if (shift > 0) {
+	new_index += current_val << shift;
```

Comment:
nit: indent

Reply:
Done - at least I see it indented. Let me know if you don't. Same for previous comment.